### PR TITLE
Build app logic

### PR DESCRIPTION
**Pull request by hpham2687**
- Add msw api for all route
- Add logic for auth 
- Add logic for add/edit/remove post
- Add logic for show posts, load more at home/manage page
- Add logic search at home page
- Add logic user can edit post at detail page

### CODE REVIEW

File: src/components/AddPost/index.js [hunk 1]
```diff
@@ -1,35 +1,30 @@
...
-    //   })
-    // );
+    await dispatch(
+      createPostsAction({
+        title,
+        content,
+        picture,
+      })
+    ).unwrap();
+    history("/manage");
```
Comment (nvh95) [inline src/components/AddPost/index.js:27]: If `createPostsAction` fails, I don't think it's a great idea to push to `/manage`

File: src/components/Banner/Banner.js [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+    setSearchText(value);
+    debouncedCallApi(value);
+  };
+
   return (
     <StyledBanner>
-      <img src="https://res.cloudinary.com/practicaldev/image/fetch/s--8G3GSc8q--/c_imagga_scale,f_auto,fl_progressive,h_420,q_auto,w_1000/https://dev-to-uploads.s3.amazonaws.com/uploads/articles/vahl9yw6xgo1a2fedl0u.jpg"></img>
+      <img
+        alt="banner"
+        src="https://res.cloudinary.com/practicaldev/image/fetch/s--8G3GSc8q--/c_imagga_scale,f_auto,fl_progressive,h_420,q_auto,w_1000/https://dev-to-uploads.s3.amazonaws.com/uploads/articles/vahl9yw6xgo1a2fedl0u.jpg"
```
Comment (nvh95) [inline src/components/Banner/Banner.js:34]: I think we can host our own images in the scope of this project

File: src/components/common/PostForm.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+      return data ? data : {};
+    }, [data]),
+  });
+
+  useEffect(() => {
+    if (!data) return;
+    reset(data);
+  }, [data, reset]);
+
+  const isHasTitlenameError = errors?.title;
```
Comment (nvh95) [inline src/components/common/PostForm.js:32]: `TitleName`

File: src/components/common/PostForm.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    </PostFormWrapper>
+  );
+}
+
+const PostFormWrapper = styled.div`
+  display: flex;
+  justify-content: center;
+  padding-top: 64px;
+  height: calc(100vh - 88px);
+  background: url(https://img.freepik.com/free-vector/hand-painted-watercolor-pastel-sky-background_23-2148902771.jpg?size=626&ext=jpg);
```
Comment (nvh95) [inline src/components/common/PostForm.js:155]: please host your own image

File: src/components/common/PostForm.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  } = useForm({
+    defaultValues: useMemo(() => {
+      return data ? data : {};
+    }, [data]),
+  });
+
+  useEffect(() => {
+    if (!data) return;
+    reset(data);
+  }, [data, reset]);
```
Comment (nvh95) [inline src/components/common/PostForm.js:30]: Why do we need to do this?

File: src/components/EditPost/index.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    fetchPostDetail();
+  }, [postId]);
+
+  const onSubmitEditPost = async (postData) => {
+    try {
+      const editedPost = await editPost({ ...postData, id: postId });
+      if (editedPost) {
+        notifyPositive({ message: "Edit post sucessfully." });
```
Comment (nvh95) [inline src/components/EditPost/index.js:32]: `sucessfully`: typo

File: src/store/reducers/postsReducer.js [hunk 7]
```diff
@@ -1,12 +1,59 @@
...
+    }
+  }
+);
+
+export const removePostAction = createAsyncThunk(
+  "posts/removePost",
+  async ({ postId }, thunkAPI) => {
+    try {
+      const response = await PostApi.removePost(postId);
+      notifyPositive({ message: `Delete post ${postId} sucessfully.` });
```
Comment (nvh95) [inline src/store/reducers/postsReducer.js:36]: `sucessfully`: typo

File: src/store/reducers/postsReducer.js [hunk 8]
```diff
@@ -1,12 +1,59 @@
...
+);
 
 export const getPostsAction = createAsyncThunk(
   "posts/getPosts",
-  async ({}, thunkAPI) => {
+  async ({ page = 1, items_per_page = 6, search = null }, thunkAPI) => {
     try {
-      const response = await UserApi.getPosts();
-      console.log(response);
+      const response = await UserApi.getPosts(page, items_per_page, search);
```
Comment (nvh95) [inline src/store/reducers/postsReducer.js:56]: Is it possible that we do not use try catch?

**Review by nvh95 — CHANGES_REQUESTED** [review]
- Form validation should be change to `onChange` mode
- Image should not be require when create a new post
- It flash when there is no post to show
- Should have `Add a new post` feature in Post list as well
- If users log out then log in, they cannot edit or delete their posts at `/`
- store should be in `store/index.js`
- Avoid using try catch when calling API
- Need to handle when api returns error
- Install code spell checker to check English grammar

File: src/App.js [hunk 9]
```diff
@@ -3,17 +3,19 @@ import { ErrorBoundary } from "react-error-boundary";
...
 import "./App.css";
+import AddPostPage from "./components/AddPost";
 import ErrorFallback from "./components/common/ErrorFallback";
 import ProtectedRoute from "./components/common/ProtectedRoute";
+import EditPostPage from "./components/EditPost";
+import Home from "./components/Home";
+import LoginPage from "./components/Login";
+import ManagePage from "./components/Manage";
+import PostDetail from "./components/PostDetail";
+import RegisterPage from "./components/Register";
```
[the remark this answers is not in the document]
Comment (nvh95) [reply src/App.js:14]: Place `App.js` in components

File: src/App.js [hunk 10]
```diff
@@ -51,14 +53,15 @@ function App() {
         <ToastContainer />
 
         <ErrorBoundary FallbackComponent={ErrorFallback}>
-          <Router>
+          <Router navigator={history}>
```
Comment (hpham2687) [inline src/App.js:56]: remove history

File: src/api/postApi.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  return axiosClient.put(`/posts/${postData.id}`, postData, {
+    headers: getHeadersWithToken(),
+  });
+}
+
+function removePost(postId) {
+  return axiosClient.delete(`/posts/${postId}`, {
+    headers: getHeadersWithToken(),
+  });
+}
```
Comment (nvh95) [inline src/api/postApi.js:23]: use fetch

File: src/api/postApi.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import { getHeadersWithToken } from "../utils/api";
+import { axiosClient } from "./axiosClient";
+
+function getPostDetail(postId) {
```
Comment (hpham2687) [inline src/api/postApi.js:4]: change axios to fetch

**Review by nvh95 — APPROVED** [review]

**Pull request merged**